sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Sum.scala

```diff
@@ -89,5 +92,11 @@ case class Sum(child: Expression) extends DeclarativeAggregate with ImplicitCast
     )
   }
 
-  override lazy val evaluateExpression: Expression = sum
+  override lazy val evaluateExpression: Expression = {
+    if (sumDataType == resultType) {
+      sum
+    } else {
+      Cast(sum, resultType)
```

Comment:
After some more thoughts, I think we should still use long as buffer type to sum long, at least by default. Adding long values is faster than adding decimal values, and we shouldn't introduce this performance regression silently.

We can have an option to use decimal as buffer type, to reduce the possibility of overflow. But it should be opt-in.

Reply:
I see. I agree. Shall we revert this change here and do this in another PR?